File: pkg/java/src/test/java/dev/openfga/language/ModelValidatorShould.java
package dev.openfga.language;

import static java.util.stream.Collectors.joining;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.junit.jupiter.params.provider.Arguments.arguments;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.openfga.language.errors.DslErrorsException;
import dev.openfga.language.errors.ModelValidationSingleError;
import dev.openfga.language.errors.ParsingError;
import dev.openfga.language.util.TestsData;
import dev.openfga.language.validation.ModelValidator;
import dev.openfga.sdk.api.model.AuthorizationModel;
import java.util.List;
import java.util.stream.Stream;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class ModelValidatorShould {

    @ParameterizedTest(name = "{0}")
    @MethodSource("dslSyntaxTestCases")
    public void verifyDslSyntax(String name, String dsl, List<ParsingError> expectedErrors, boolean skip) {
        Assumptions.assumeFalse(skip);

        var thrown = catchThrowable(() -> ModelValidator.validateDsl(dsl));

        if (expectedErrors.isEmpty()) {
            assertThat(thrown).isNull();
            return;
        }

        assertThat(thrown).isInstanceOf(DslErrorsException.class);

        // unfortunately antlr is throwing different error messages in Java, Go and JS - considering that at the moment
        //  we care that it errors for syntax errors more than we care about the error messages matching,
        //  esp. in Java as we are not building a language server on top of the returned errors yet
        //  actual matching error strings is safe to ignore for now

        //        var errorsCount = expectedErrors.size();
        //
        //        var formattedErrors = expectedErrors.stream()
        //                .map(error -> String.format("syntax error at line=%d, column=%d: %s",
        // error.getLine().getStart(), error.getColumn().getStart(), error.getMessage()))
        //                .collect(joining("\n\t* "));
        //
        //        var expectedMessage = String.format("%d error%s occurred:\n\t* %s\n\n",
        //                errorsCount,
        //                errorsCount > 1 ? "s" : "",
        //                formattedErrors);
        //
        //        assertThat(thrown).hasMessage(expectedMessage);
        //
        //        var actualErrors = ((DslErrorsException) thrown).getErrors();
        //        for (int i = 0; i < expectedErrors.size(); i++) {
        //            var expectedError = expectedErrors.get(i);
        //            var actualError = actualErrors.get(i);
        //
        //            assertMatch(expectedError, actualError);
        //        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("dslValidationTestCases")
    public void verifyDslValidation(
            String name, String dsl, List<ModelValidationSingleError> expectedErrors, boolean skip) {
        Assumptions.assumeFalse(skip);

        var thrown = catchThrowable(() -> ModelValidator.validateDsl(dsl));

        if (expectedErrors.isEmpty()) {
            assertThat(thrown).isNull();
            return;
        }

        assertThat(thrown).isInstanceOf(DslErrorsException.class);

        var errorsCount = expectedErrors.size();

        var formattedErrors = expectedErrors.stream()
                .map(error -> String.format(
                        "validation error at line=%d, column=%d: %s",
                        error.getLine().getStart(), error.getColumn().getStart(), error.getMessage()))
                .collect(joining("\n\t* "));

        var expectedMessage = String.format(
                "%d error%s occurred:\n\t* %s\n\n", errorsCount, errorsCount > 1 ? "s" : "", formattedErrors);

        assertThat(thrown).hasMessage(expectedMessage);

        var actualErrors = ((DslErrorsException) thrown).getErrors();
        for (int i = 0; i < expectedErrors.size(); i++) {
            var expectedError = expectedErrors.get(i);
            var actualError = actualErrors.get(i);

            assertMatch(expectedError, (ModelValidationSingleError) actualError);
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("jsonValidationTestCases")
    public void verifyJsonValidation(String name, String json, List<ModelValidationSingleError> expectedErrors)
            throws JsonProcessingException {
        var model = new ObjectMapper().readValue(json, AuthorizationModel.class);

        var thrown = catchThrowable(() -> ModelValidator.validateJson(model));

        if (expectedErrors == null || expectedErrors.isEmpty()) {
            assertThat(thrown).isNull();
            return;
        }

        assertThat(thrown).isInstanceOf(DslErrorsException.class);

        var errorsCount = expectedErrors.size();

        var formattedErrors = expectedErrors.stream()
                .map(error -> String.format("validation error: %s", error.getMessage()))
                .collect(joining("\n\t* "));

        var expectedMessage = String.format(
                "%d error%s occurred:\n\t* %s\n\n", errorsCount, errorsCount > 1 ? "s" : "", formattedErrors);

        assertThat(thrown).hasMessage(expectedMessage);

        var actualErrors = ((DslErrorsException) thrown).getErrors();
        for (int i = 0; i < expectedErrors.size(); i++) {
            var expectedError = expectedErrors.get(i);
            var actualError = actualErrors.get(i);

            assertMatch(expectedError, (ModelValidationSingleError) actualError);
        }
    }

    private void assertMatch(ParsingError expectedError, ParsingError actualError) {
        assertThat(actualError.getMessage()).isEqualTo(expectedError.getMessage());
        assertThat(actualError.getLine()).isEqualTo(expectedError.getLine());
        assertThat(actualError.getColumn()).isEqualTo(expectedError.getColumn());
    }

    private void assertMatch(ModelValidationSingleError expectedError, ModelValidationSingleError actualError) {
        assertThat(actualError.getMessage()).isEqualTo(expectedError.getMessage());

        if (expectedError.getLine() != null) {
            assertThat(actualError.getLine()).isEqualTo(expectedError.getLine());
        }

        if (expectedError.getColumn() != null) {
            assertThat(actualError.getColumn()).isEqualTo(expectedError.getColumn());
        }

        assertThat(actualError.getMetadata().getErrorType())
                .isEqualTo(expectedError.getMetadata().getErrorType());
        if (expectedError.getMetadata().getTypeName() != null) {
            assertThat(actualError.getMetadata().getTypeName())
                    .isEqualTo(expectedError.getMetadata().getTypeName());
        }
        if (expectedError.getMetadata().getRelation() != null) {
            assertThat(actualError.getMetadata().getRelation())
                    .isEqualTo(expectedError.getMetadata().getRelation());
        }
        if (expectedError.getMetadata().getConditionName() != null) {
            assertThat(actualError.getMetadata().getConditionName())
                    .isEqualTo(expectedError.getMetadata().getConditionName());
        }
    }

    private static Stream<Arguments> dslSyntaxTestCases() {
        return TestsData.DSL_SYNTAX_TEST_CASES.stream()
                .map(testCase -> arguments(
                        testCase.getName(), testCase.getDsl(), testCase.getExpectedErrors(), testCase.isSkip()));
    }

    private static Stream<Arguments> dslValidationTestCases() {
        return TestsData.DSL_VALIDATION_TEST_CASES.stream()
                .map(testCase -> arguments(
                        testCase.getName(), testCase.getDsl(), testCase.getExpectedErrors(), testCase.isSkip()));
    }

    private static Stream<Arguments> jsonValidationTestCases() {
        return TestsData.JSON_VALIDATION_TEST_CASES.stream()
                .map(testCase -> arguments(testCase.getName(), testCase.getJson(), testCase.getExpectedErrors()));
    }
}

File: pkg/csharp/test/OpenFga.Language.Tests/ModelValidatorShould.cs
#pragma warning disable IDE0060
#pragma warning disable xUnit1026

using OpenFga.Language.Errors;
using OpenFga.Language.Tests.util;
using OpenFga.Language.Validation;
using OpenFga.Sdk.Model;
using System.Text.Json;
using Xunit;

namespace OpenFga.Language.Tests;

public class ModelValidatorShould {
    [Theory]
    [MemberData(nameof(DslSyntaxTestCases))]
    public void VerifyDslSyntax(string name, string dsl, List<ModelValidationSingleError> expectedErrors, bool skip) {
        if (skip) {
            return;
        }

        if (expectedErrors.Count == 0) {
            // If no errors expected, validation should succeed
            ModelValidator.ValidateDsl(dsl);
            return;
        }

        // If errors are expected, validation should throw DslErrorsException
        var exception = Assert.Throws<DslErrorsException>(() => ModelValidator.ValidateDsl(dsl));

        // Unfortunately antlr is throwing different error messages in Java, Go and JS - considering that at the moment
        // we care that it errors for syntax errors more than we care about the error messages matching,
        // esp. in Java as we are not building a language server on top of the returned errors yet
        // actual matching error strings is safe to ignore for now

        // var errorsCount = expectedErrors.Count;
        // var formattedErrors = expectedErrors
        //     .Select(error => $"syntax error at line={error.Line?.Start}, column={error.Column?.Start}: {error.Message}")
        //     .ToList();
        // var expectedMessage = $"{errorsCount} error{(errorsCount > 1 ? "s" : "")} occurred:\n\t* {string.Join("\n\t* ", formattedErrors)}\n\n";
        // Assert.Equal(expectedMessage, exception.Message);
        // var actualErrors = exception.Errors;
        // for (int i = 0; i < expectedErrors.Count; i++)
        // {
        //     var expectedError = expectedErrors[i];
        //     var actualError = actualErrors[i];
        //     AssertMatch(expectedError, actualError);
        // }
    }

    [Theory]
    [MemberData(nameof(DslValidationTestCases))]
    public void VerifyDslValidation(
        string name, string dsl, List<ModelValidationSingleError> expectedErrors, bool skip) {
        if (skip) {
            return;
        }

        if (expectedErrors.Count == 0) {
            // If no errors expected, validation should succeed
            ModelValidator.ValidateDsl(dsl);
            return;
        }

        // If errors are expected, validation should throw DslErrorsException
        var exception = Assert.Throws<DslErrorsException>(() => ModelValidator.ValidateDsl(dsl));

        var errorsCount = expectedErrors.Count;
        var formattedErrors = expectedErrors
            .Select(error => $"validation error at line={error.Line?.Start}, column={error.Column?.Start}: {error.Message}")
            .ToList();
        var expectedMessage = $"{errorsCount} error{(errorsCount > 1 ? "s" : "")} occurred:\n\t* {string.Join("\n\t* ", formattedErrors)}\n\n";

        Assert.Equal(expectedMessage, exception.Message);

        var actualErrors = exception.Errors;
        for (int i = 0; i < expectedErrors.Count; i++) {
            var expectedError = expectedErrors[i];
            var actualError = actualErrors[i];

            AssertMatch(expectedError, (ModelValidationSingleError)actualError);
        }
    }

    [Theory]
    [MemberData(nameof(JsonValidationTestCases))]
    public void VerifyJsonValidation(string name, string json, List<ModelValidationSingleError>? expectedErrors) {
        var model = JsonSerializer.Deserialize<AuthorizationModel>(json)!;

        if (expectedErrors == null || expectedErrors.Count == 0) {
            // If no errors expected, validation should succeed
            ModelValidator.ValidateJson(model);
            return;
        }

        // If errors are expected, validation should throw DslErrorsException
        var exception = Assert.Throws<DslErrorsException>(() => ModelValidator.ValidateJson(model));

        var errorsCount = expectedErrors.Count;
        var formattedErrors = expectedErrors
            .Select(error => $"validation error: {error.Message}")
            .ToList();
        var expectedMessage = $"{errorsCount} error{(errorsCount > 1 ? "s" : "")} occurred:\n\t* {string.Join("\n\t* ", formattedErrors)}\n\n";

        Assert.Equal(expectedMessage, exception.Message);

        var actualErrors = exception.Errors;
        for (int i = 0; i < expectedErrors.Count; i++) {
            var expectedError = expectedErrors[i];
            var actualError = actualErrors[i];

            AssertMatch(expectedError, (ModelValidationSingleError)actualError);
        }
    }

    private void AssertMatch(ParsingError expectedError, ParsingError actualError) {
        Assert.Equal(expectedError.Message, actualError.Message);
        Assert.Equal(expectedError.Line, actualError.Line);
        Assert.Equal(expectedError.Column, actualError.Column);
    }

    private void AssertMatch(ModelValidationSingleError expectedError, ModelValidationSingleError actualError) {
        Assert.Equal(expectedError.Message, actualError.Message);

        if (expectedError.Line != null) {
            Assert.Equal(expectedError.Line, actualError.Line);
        }

        if (expectedError.Column != null) {
            Assert.Equal(expectedError.Column, actualError.Column);
        }

        Assert.Equal(expectedError.Metadata.ErrorType, actualError.Metadata.ErrorType);
        if (expectedError.Metadata.TypeName != null) {
            Assert.Equal(expectedError.Metadata.TypeName, actualError.Metadata.TypeName);
        }
        if (expectedError.Metadata.Relation != null) {
            Assert.Equal(expectedError.Metadata.Relation, actualError.Metadata.Relation);
        }
        if (expectedError.Metadata.Condition != null) {
            Assert.Equal(expectedError.Metadata.Condition, actualError.Metadata.Condition);
        }
    }

    public static IEnumerable<object[]> DslSyntaxTestCases() {
        return TestsData.DslSyntaxTestCases
            .Select(testCase => new object[]
            {
                testCase.Name,
                testCase.Dsl,
                testCase.ExpectedErrors,
                testCase.Skip
            });
    }

    public static IEnumerable<object[]> DslValidationTestCases() {
        return TestsData.DslValidationTestCases
            .Select(testCase => new object[]
            {
                testCase.Name,
                testCase.Dsl,
                testCase.ExpectedErrors,
                testCase.Skip
            });
    }

    public static IEnumerable<object[]> JsonValidationTestCases() {
        return TestsData.JsonValidationTestCases
            .Select(testCase => new object[]
            {
                testCase.Name,
                testCase.Json,
                testCase.ExpectedErrors!
            });
    }
}